src/qt/trafficgraphwidget.h

```diff
@@ -1,48 +1,63 @@
-// Copyright (c) 2011-2015 The Bitcoin Core developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#ifndef BITCOIN_QT_TRAFFICGRAPHWIDGET_H
-#define BITCOIN_QT_TRAFFICGRAPHWIDGET_H
-
-#include <QWidget>
-#include <QQueue>
-
-class ClientModel;
-
-QT_BEGIN_NAMESPACE
-class QPaintEvent;
-class QTimer;
-QT_END_NAMESPACE
-
-class TrafficGraphWidget : public QWidget
-{
-    Q_OBJECT
-
-public:
-    explicit TrafficGraphWidget(QWidget *parent = 0);
-    void setClientModel(ClientModel *model);
-    int getGraphRangeMins() const;
-
-protected:
-    void paintEvent(QPaintEvent *);
-
-public Q_SLOTS:
-    void updateRates();
-    void setGraphRangeMins(int mins);
-    void clear();
-
-private:
-    void paintPath(QPainterPath &path, QQueue<float> &samples);
-
-    QTimer *timer;
-    float fMax;
-    int nMins;
-    QQueue<float> vSamplesIn;
-    QQueue<float> vSamplesOut;
-    quint64 nLastBytesIn;
-    quint64 nLastBytesOut;
-    ClientModel *clientModel;
-};
-
-#endif // BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+// Copyright (c) 2011-2015 The Bitcoin Core developers
+// Copyright (c) 2022 The Dogecoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+#define BITCOIN_QT_TRAFFICGRAPHWIDGET_H
+
+#include <QWidget>
+#include <QVector>
+
+class ClientModel;
+
+QT_BEGIN_NAMESPACE
+class QPaintEvent;
+class QTimer;
+QT_END_NAMESPACE
+
+class TrafficGraphWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    explicit TrafficGraphWidget(QWidget *parent = 0);
+    void setClientModel(ClientModel *model);
+    int getGraphRangeMins() const;
+
+protected:
+    void paintEvent(QPaintEvent *);
+
+public Q_SLOTS:
+    void updateTimespanRates();
+    void updateBaseRates();
+    void setGraphRangeMins(int mins);
+    void clear();
+
+private:
+    void paintPath(QPainterPath &path, const QVector<float> &samples, int samplesHead);
+    float resampleSamples(const QVector<float> &source, QVector<float> &dest) const;
+
+    float fMax;
+    int nMins;
+
+    QVector<float> vTimespanSamplesIn;
+    QVector<float> vTimespanSamplesOut;
+    quint64 nTimespanLastBytesIn;
+    quint64 nTimespanLastBytesOut;
+    int64_t nTimespanLastSampleTimeMs;
+    int nTimespanHeadIndex;
+    QTimer *timespanTimer;
+
+    QVector<float> vBaseSamplesIn;
+    QVector<float> vBaseSamplesOut;
+    quint64 nBaseLastBytesIn;
+    quint64 nBaseLastBytesOut;
+    int64_t nBaseLastSampleTimeMs;
+    int nBaseSamplesHeadIndex;
+    QTimer *baseTimer;
+
+    ClientModel *clientModel;
+};
+
+#endif // BITCOIN_QT_TRAFFICGRAPHWIDGET_H
```

Comment:
Nit: need to bring back trailing newline

Reply:
fixed